src/mem/heap/heap_simple.c

```diff
@@ -1,116 +1,132 @@
 /**
  * @file
- * @brief Pseudo-dynamic memory allocator.
+ * @brief Simple memory allocator
  *
  * @date 13.10.09
- * @author Nikolay Korotky
+ * @author Nikolay Korotky -- initial implementation
+ * @author Denis Deryugin -- rework
  */
 
-
 #include <string.h>
 #include <stdlib.h>
 #include <embox/unit.h>
 #include <mem/heap.h>
 #include <mem/page.h>
+#include <kernel/printk.h>
+#include <util/binalign.h>
+#include <util/log.h>
 
-static int has_initialized = 0;
-static char *managed_memory_start;
-static char *last_valid_address;
+static int simple_heap_intialized = 0;
+static void *heap_simple_start;
+static void *heap_simple_end;
 
 #define HEAP_SIZE OPTION_MODULE_GET(embox__mem__heap_api,NUMBER,heap_size)
+#define HEAP_ALIGN OPTION_MODULE_GET(embox__mem__heap_api,NUMBER,align)
 
 #define MEM_POOL_SIZE  (HEAP_SIZE / PAGE_SIZE())
 static void *mem_pool;
 
 struct mem_control_block {
 	int is_available;
-	size_t size;
+	size_t size; /* Length of memory chunk WITHOUT this structure
+	              * (which is located at the beginning of the chunk) */
 };
 
+#define MCB_SZ binalign_bound(sizeof(struct mem_control_block), HEAP_ALIGN)
+
 static int malloc_init(void) {
 	extern struct page_allocator *__heap_pgallocator;
 	struct mem_control_block *init_mcb;
 
 	mem_pool = page_alloc(__heap_pgallocator, MEM_POOL_SIZE);
-	if(NULL == mem_pool) {
+	if (NULL == mem_pool) {
 		return -1;
 	}
 
-	managed_memory_start = (void*) mem_pool;
-	last_valid_address   = managed_memory_start;
-	has_initialized      = 1;
+	heap_simple_start =
+		(void*) binalign_bound((uintptr_t) mem_pool, HEAP_ALIGN);
```

Comment:
I think we should assert heap_simple_start is already aligned to HEAP_ALIGN bound after `page_alloc` above since if it's not so, we will align get out of bounds below 

heap_simple_end   = heap_simple_start + HEAP_SIZE;